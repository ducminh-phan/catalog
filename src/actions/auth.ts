import { createAction, createAsyncThunk } from "@reduxjs/toolkit";

import Request from "utils/api";

export interface LoginRequest {
  username: string;
  password: string;
}

export interface LoginResponse {
  access_token: string;
}

export const login = createAsyncThunk("auth/login", (body: LoginRequest) => {
  return Request.post<LoginResponse>("login", body);
});

export interface RegisterRequest {
  username: string;
  password: string;
  email: string;
  name: string;
}

export interface RegisterResponse {
  access_token: string;
}

export const register = createAsyncThunk(
  "auth/register",
  (body: RegisterRequest) =>
    Request.post<RegisterResponse>("registrations", body),
);

export const logout = createAction("auth/logout");
