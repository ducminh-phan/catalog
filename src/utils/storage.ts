class Storage {
  private readonly prefix: string;

  private static readonly tokenKey: "token";

  constructor(localStoragePrefix: string) {
    this.prefix = localStoragePrefix;
  }

  getItem = (key: string): string | null =>
    localStorage.getItem(`${this.prefix}:${key}`);

  setItem = (key: string, value: string): void =>
    localStorage.setItem(`${this.prefix}:${key}`, value);

  removeItem = (key: string): void =>
    localStorage.removeItem(`${this.prefix}:${key}`);

  getToken = (): string | null => this.getItem(Storage.tokenKey);

  setToken = (token: string): void => this.setItem(Storage.tokenKey, token);

  removeToken = (): void => this.removeItem(Storage.tokenKey);
}

export default new Storage(process.env.REACT_APP_LOCAL_STORAGE_PREFIX);
