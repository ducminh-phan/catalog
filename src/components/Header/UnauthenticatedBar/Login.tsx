import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  TextField,
} from "@material-ui/core";
import { createAction, createReducer, PrepareAction } from "@reduxjs/toolkit";
import React, { ChangeEvent } from "react";
import { connect, ConnectedProps } from "react-redux";

import { login } from "actions/auth";

interface ChangeInputPayload {
  name: string;
  value: string;
}

const changeInput = createAction<PrepareAction<ChangeInputPayload>>(
  "login/changeInput",
  ({ target: { name, value } }: ChangeEvent<HTMLInputElement>) => ({
    payload: {
      name,
      value,
    },
  }),
);

const initialState = {
  username: "",
  password: "",
};

const reducer = createReducer(initialState, (builder) =>
  builder.addCase(changeInput, (state, action) => {
    const { name, value } = action.payload;

    return {
      ...state,
      [name]: value,
    };
  }),
);

const connector = connect(null, { login });
type PropsFromRedux = ConnectedProps<typeof connector>;
type Props = PropsFromRedux;

const Login = (props: Props): React.ReactElement => {
  const [open, setOpen] = React.useState(false);

  const handleOpen = (): void => {
    setOpen(true);
  };

  const handleClose = (): void => {
    setOpen(false);
  };

  const [{ username, password }, dispatch] = React.useReducer(
    reducer,
    initialState,
  );

  const handleLogin = (): void => {
    setOpen(false);
    props.login({ username, password });
  };

  return (
    <>
      <Button color="inherit" onClick={handleOpen}>
        Login
      </Button>
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="register-form-title"
        fullWidth
      >
        <DialogTitle id="register-form-title">Login</DialogTitle>
        <DialogContent>
          <TextField
            autoFocus
            label="Username"
            onChange={(e): void => {
              dispatch(changeInput(e));
            }}
            name="username"
            value={username}
            required
            fullWidth
          />
          <TextField
            label="Password"
            type="password"
            onChange={(e): void => {
              dispatch(changeInput(e));
            }}
            name="password"
            value={password}
            required
            fullWidth
          />
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose} color="primary">
            Cancel
          </Button>
          <Button onClick={handleLogin} color="primary">
            Login
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

export default connector(Login);
