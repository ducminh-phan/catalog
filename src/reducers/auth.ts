import { createReducer } from "@reduxjs/toolkit";

import { login } from "actions/auth";
import storage from "utils/storage";

interface AuthState {
  token: string | null;
}

const initialState: AuthState = {
  token: storage.getToken(),
};

export default createReducer(initialState, (builder) =>
  builder.addCase(
    login.fulfilled,
    (state, { payload: { access_token: token } }) => ({ token }),
  ),
);
