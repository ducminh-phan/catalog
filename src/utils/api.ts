import { Json, snakeToCamel } from "./misc";
import storage from "./storage";

type Method = "get" | "post" | "put";

class Request {
  static request = async <T extends Json>(
    endpoint: string,
    method: Method,
    body?: object,
  ): Promise<T> => {
    const token = storage.getToken();
    const headers: HeadersInit = {
      Accept: "application/json",
    };

    if (token) {
      headers.Authorization = `Bearer ${token}`;
    }

    const config: RequestInit = {
      method,
      headers,
    };

    if (body) {
      config.body = JSON.stringify(body);
      headers["Content-Type"] = "application/json";
    }

    const response = await window.fetch(
      `${process.env.REACT_APP_API_URL}/${endpoint}`,
      config,
    );

    const data = snakeToCamel(await response.json());

    if (response.ok) {
      return data as T;
    }

    return Promise.reject(data);
  };

  static get = <T extends Json>(endpoint: string): Promise<T> =>
    Request.request(endpoint, "get");

  static post = <T extends Json>(endpoint: string, body: object): Promise<T> =>
    Request.request<T>(endpoint, "post", body);

  static put = <T extends Json>(endpoint: string, body: object): Promise<T> =>
    Request.request<T>(endpoint, "put", body);
}

export default Request;
