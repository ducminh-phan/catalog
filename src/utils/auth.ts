import Request from "./api";
import storage from "./storage";

export type LoginRequest = {
  username: string;
  password: string;
};

export type LoginResponse = {
  accessToken: string;
};

export type RegisterRequest = {
  username: string;
  password: string;
  email: string;
  name: string;
};

export type RegisterResponse = {
  accessToken: string;
};

const handleAuthResponse = ({
  accessToken,
}: LoginResponse | RegisterResponse): void => storage.setToken(accessToken);

export const login = (body: LoginRequest): Promise<void> => {
  return Request.post<LoginResponse>("login", body).then(handleAuthResponse);
};

export const register = (body: RegisterRequest): Promise<void> => {
  return Request.post<RegisterResponse>("registrations", body).then(
    handleAuthResponse,
  );
};

export const logout = (): Promise<void> => {
  storage.removeToken();
  return Promise.resolve();
};

interface User {
  id: number;
  name: string;
}

export interface AppData {
  user: User;
}

type UserResponse = {
  email: string;
  username: string;
  created: string;
  updated: string;
  name: string;
  id: number;
};

export const getAppData = async (): Promise<AppData | null> => {
  const token = storage.getToken();

  if (token === null) {
    return null;
  }

  return Request.get<UserResponse>("me")
    .then(({ id, name }) => ({ user: { id, name } }))
    .catch((error) => {
      logout();
      return Promise.reject(error);
    });
};
