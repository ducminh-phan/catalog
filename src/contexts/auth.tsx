import React, {
  createContext,
  PropsWithChildren,
  ReactElement,
  useContext,
} from "react";
import { useAsync } from "react-async";

import * as auth from "utils/auth";
import storage from "utils/storage";

import { useNotification } from "./notification";

interface AuthContextValue {
  data: auth.AppData | null;
  register: (payload: auth.RegisterRequest) => void;
  login: (payload: auth.LoginRequest) => void;
  logout: () => void;
}

const AuthContext = createContext<AuthContextValue | undefined>(undefined);

export const AuthProvider = (
  props: PropsWithChildren<object>,
): ReactElement => {
  const [firstAttemptFinished, setFirstAttemptFinished] = React.useState(false);
  const { data = null, error, isRejected, isSettled, reload } = useAsync({
    promiseFn: auth.getAppData,
  });

  React.useLayoutEffect(() => {
    if (isSettled) {
      setFirstAttemptFinished(true);
    }
  }, [isSettled]);

  const { setNotification } = useNotification();

  if (!firstAttemptFinished) {
    if (isRejected) {
      setNotification({
        severity: "error",
        message: error?.message ?? "Something went wrong",
      });
    }

    return <div />;
  }

  const login = (payload: auth.LoginRequest): void => {
    auth
      .login(payload)
      .then(reload)
      .then(() =>
        Promise.resolve(
          setNotification({
            severity: "success",
            message: "Login successfully",
          }),
        ),
      )
      .catch(({ message }) => {
        setNotification({ severity: "error", message });
      });
  };

  const register = (payload: auth.RegisterRequest): void => {
    auth
      .register(payload)
      .then(reload)
      .then(() =>
        Promise.resolve(
          setNotification({
            severity: "success",
            message: "Register successfully",
          }),
        ),
      )
      .catch(({ message }) => {
        setNotification({
          severity: "error",
          message: Object.values(message)[0] as string,
        });
      });
  };

  const logout = (): void => {
    auth
      .logout()
      .then(() =>
        Promise.resolve(
          setNotification({
            severity: "success",
            message: "Logout successfully",
          }),
        ),
      )
      .then(reload);
  };

  return (
    <AuthContext.Provider
      value={{ data, login, logout, register }}
      /* eslint-disable-next-line react/jsx-props-no-spreading */
      {...props}
    />
  );
};

export const useAuth = (): AuthContextValue => {
  const context = useContext<AuthContextValue | undefined>(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within a AuthProvider");
  }
  return context;
};
