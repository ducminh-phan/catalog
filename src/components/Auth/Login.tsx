import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  TextField,
} from "@material-ui/core";
import { createAction, createReducer, PrepareAction } from "@reduxjs/toolkit";
import React, { ChangeEvent } from "react";

import { useAuth } from "contexts/auth";

interface ChangeInputPayload {
  name: string;
  value: string;
}

const changeInput = createAction<PrepareAction<ChangeInputPayload>>(
  "login/changeInput",
  ({ target: { name, value } }: ChangeEvent<HTMLInputElement>) => ({
    payload: {
      name,
      value,
    },
  }),
);

const initialState = {
  username: "",
  password: "",
};

const reducer = createReducer(initialState, (builder) =>
  builder.addCase(changeInput, (state, action) => {
    const { name, value } = action.payload;

    return {
      ...state,
      [name]: value,
    };
  }),
);

const Login = (): React.ReactElement => {
  const [open, setOpen] = React.useState(false);

  const handleOpen = (): void => {
    setOpen(true);
  };

  const handleClose = (): void => {
    setOpen(false);
  };

  const [{ username, password }, dispatch] = React.useReducer(
    reducer,
    initialState,
  );

  const { login } = useAuth();

  const handleLogin = (): void => {
    login({ username, password });
  };

  return (
    <>
      <Button color="inherit" onClick={handleOpen}>
        Login
      </Button>
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="register-form-title"
        fullWidth
      >
        <DialogTitle id="register-form-title">Login</DialogTitle>
        <DialogContent>
          <TextField
            autoFocus
            label="Username"
            onChange={(e): void => {
              dispatch(changeInput(e));
            }}
            name="username"
            value={username}
            required
            fullWidth
          />
          <TextField
            label="Password"
            type="password"
            onChange={(e): void => {
              dispatch(changeInput(e));
            }}
            name="password"
            value={password}
            required
            fullWidth
          />
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose} color="primary">
            Cancel
          </Button>
          <Button
            onClick={handleLogin}
            color="primary"
            disabled={!!(!username || !password)}
          >
            Login
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

export default Login;
