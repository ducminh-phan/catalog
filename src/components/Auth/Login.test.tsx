import React from "react";

import Request from "utils/api";
import { LoginResponse } from "utils/auth";
import storage from "utils/storage";
import {
  fireEvent,
  render,
  wait,
  waitForElementToBeRemoved,
} from "utils/tests";

import Login from "./Login";

it("can login", async () => {
  const {
    baseElement,
    getByLabelText,
    getByRole,
    getByText,
    queryByText,
  } = render(<Login />);

  await wait(() => getByText("Login"));
  expect(queryByText("Login")).toBeInTheDocument();
  expect(queryByText("Cancel")).toBeNull();

  // Click the Login button
  fireEvent.click(getByText("Login"));
  await wait(() => getByText("Cancel"));
  expect(queryByText("Password")).toBeInTheDocument();
  expect(queryByText("Username")).toBeInTheDocument();
  expect(baseElement).toMatchSnapshot();

  // Click the Cancel button
  fireEvent.click(getByText("Cancel"));
  await waitForElementToBeRemoved(() => queryByText("Cancel"));
  expect(queryByText("Password")).toBeNull();
  expect(queryByText("Username")).toBeNull();

  // Click the Login button again
  fireEvent.click(getByText("Login"));
  await wait(() => getByText("Cancel"));

  // The login button should be disabled
  const loginButton = [
    ...getByRole("presentation").querySelectorAll("button"),
  ].filter((node) => node.textContent === "Login")[0];
  expect(loginButton).toHaveAttribute("disabled");

  // Fill in mock username and password
  const usernameField = getByLabelText(/Username.*/) as HTMLInputElement;
  const passwordField = getByLabelText(/Password.*/) as HTMLInputElement;
  const mockUsername = "mockUsername";
  const mockPassword = "mockPassword";
  await wait(() => {
    fireEvent.change(usernameField, {
      target: { value: mockUsername },
    });
    fireEvent.change(passwordField, {
      target: { value: mockPassword },
    });
  });

  expect(usernameField.value).toBe(mockUsername);
  expect(passwordField.value).toBe(mockPassword);
  expect(loginButton).not.toHaveAttribute("disabled");

  // Login
  jest
    .spyOn(Request, "post")
    .mockImplementation(
      (): Promise<LoginResponse> =>
        Promise.resolve({ accessToken: "accessToken" }),
    );

  await wait(() => fireEvent.click(loginButton));
  expect(Request.post).toHaveBeenCalledWith("login", {
    password: mockPassword,
    username: mockUsername,
  });
  expect(storage.getToken()).toBe("accessToken");
});
