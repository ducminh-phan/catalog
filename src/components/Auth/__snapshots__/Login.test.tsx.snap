// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can log in 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Login
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby="login-form-title"
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
          id="login-form-title"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            Login
          </h2>
        </div>
        <div
          class="MuiDialogContent-root"
        >
          <form
            novalidate=""
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiFormLabel-required MuiInputLabel-required"
                data-shrink="false"
                for="login-username"
                id="login-username-label"
              >
                Username
                <span
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="login-username"
                  name="username"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiFormLabel-required MuiInputLabel-required"
                data-shrink="false"
                for="login-password"
                id="login-password-label"
              >
                Password
                <span
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="login-password"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              class="MuiDialogActions-root MuiDialogActions-spacing"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Cancel
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-disabled MuiButtonBase-disabled"
                disabled=""
                tabindex="-1"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                  Login
                </span>
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
